src/crypto/ocb.js

```diff
@@ -0,0 +1,332 @@
+ */
+function double(data) {
+  const double = new Uint8Array(16);
+  for (let i = 0; i < 15; i++) {
+    double[i] = (data[i] << 1) ^ (data[i + 1] >> 7);
+  }
+  double[15] = (data[15] << 1) ^ ((data[0] >> 7) * 0x87);
+  return double;
+}
+
+
+const zeroBlock = new Uint8Array(blockLength);
+const one = new Uint8Array([1]);
+
+/**
+ * Class to en/decrypt using OCB mode.
+ * @param  {String}     cipher      The symmetric cipher algorithm to use e.g. 'aes128'
+ * @param  {Uint8Array} key         The encryption key
+ */
+async function OCB(cipher, key) {
+
+  let maxNtz = 0;
+  let kv;
+
+  constructKeyVariables(cipher, key);
+
+  function constructKeyVariables(cipher, key) {
+    const aes = new ciphers[cipher](key);
+    const encipher = aes.encrypt.bind(aes);
+    const decipher = aes.decrypt.bind(aes);
+
+    const mask_x = encipher(zeroBlock);
+    const mask_$ = double(mask_x);
+    const mask = [];
+    mask[0] = double(mask_$);
+
+
+    mask.x = mask_x;
+    mask.$ = mask_$;
+
+    kv = { encipher, decipher, mask };
+  }
+
+  function extendKeyVariables(text, adata) {
+    const { mask } = kv;
+    const newMaxNtz = util.nbits(Math.max(text.length, adata.length) >> 4) - 1;
+    for (let i = maxNtz + 1; i <= newMaxNtz; i++) {
+      mask[i] = double(mask[i - 1]);
+    }
+    maxNtz = newMaxNtz;
+  }
+
+  function hash(adata) {
+    if (!adata.length) {
+      // Fast path
+      return zeroBlock;
+    }
+
+    const { encipher, mask } = kv;
+
+    //
+    // Consider A as a sequence of 128-bit blocks
+    //
+    const m = adata.length >> 4;
+
+    const offset = new Uint8Array(16);
+    const sum = new Uint8Array(16);
+    for (let i = 0; i < m; i++) {
+      xorMut(offset, mask[ntz(i + 1)]);
+      xorMut(sum, encipher(xor(offset, adata)));
+      adata = adata.subarray(16);
+    }
+
+    //
+    // Process any final partial block; compute final hash value
+    //
+    if (adata.length) {
+      xorMut(offset, mask.x);
+
+      const cipherInput = new Uint8Array(16);
+      cipherInput.set(adata, 0);
+      cipherInput[adata.length] = 0b10000000;
+      xorMut(cipherInput, offset);
+
+      xorMut(sum, encipher(cipherInput));
+    }
+
+    return sum;
+  }
+
+
+  return {
+    /**
+     * Encrypt plaintext input.
+     * @param  {Uint8Array} plaintext   The cleartext input to be encrypted
+     * @param  {Uint8Array} nonce       The nonce (15 bytes)
+     * @param  {Uint8Array} adata       Associated data to sign
+     * @returns {Promise<Uint8Array>}    The ciphertext output
+     */
+    encrypt: async function(plaintext, nonce, adata) {
+      //
+      // Consider P as a sequence of 128-bit blocks
+      //
+      const m = plaintext.length >> 4;
+
+      //
+      // Key-dependent variables
+      //
+      extendKeyVariables(plaintext, adata);
+      const { encipher, mask } = kv;
+
+      //
+      // Nonce-dependent and per-encryption variables
+      //
+      // We assume here that TAGLEN mod 128 == 0 (tagLength === 16).
```

Comment:
do you mean 128 mod 16 = 0?

Reply:
`TAGLEN mod 128` is a quote from the paper, in which `TAGLEN` is any number of bits up to 128. I.e. we assume `tagLength mod 16 == 0`, i.e. `tagLength == 16` (I guess the paper technically allows `tagLength == 0`, too).

Reply:
Ohh got it, yeah I would just change this to `tagLength mod 16` to avoid confusion